### ISSUE TITLE
Dead Code Elimination Regression at -Oz (trunk vs. 13.0.0)

### ISSUE THREAD

**Issue by Quarub** [llvm/llvm-project#53321]
`cat case.c #13586`
[case.txt](https://github.com/llvm/llvm-project/files/7907437/case.txt)
```c
unsigned b;
void foo();
static short(a)(unsigned short c, int d) { return c > 63488 >> d; }
void e(int c) { b = a(c, 11); }
int main() {
  char f = 0;
  for (; f < 8; f++)
    e(f);
  if (b)
    foo();
}
```
`clang-f29256a64ac11cf59cea878c8d1ba9537db4f523 (trunk) -Oz` can not eliminate `foo` but `clang-llvmorg-13.0.0 -Oz` can.

Target: `x86_64-unknown-linux-gnu`

------------------------------------------------

`clang-f29256a64ac11cf59cea878c8d1ba9537db4f523 (trunk) -Oz [-emit-llvm] -S -o /dev/stdout case.c`
<details ><summary>Emitted IR</summary><p>

```ll
; ModuleID = 'case.c'
source_filename = "case.c"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

@b = dso_local local_unnamed_addr global i32 0, align 4

; Function Attrs: minsize mustprogress nofree norecurse nosync nounwind optsize uwtable willreturn writeonly
define dso_local void @e(i32 noundef %0) local_unnamed_addr #0 {
  %2 = and i32 %0, 65504
  %3 = icmp ne i32 %2, 0
  %4 = zext i1 %3 to i32
  store i32 %4, i32* @b, align 4, !tbaa !3
  ret void
}

; Function Attrs: minsize nounwind optsize uwtable
define dso_local i32 @main() local_unnamed_addr #1 {
  %1 = load i32, i32* @b, align 4, !tbaa !3
  br label %2

2:                                                ; preds = %6, %0
  %3 = phi i32 [ %1, %0 ], [ %9, %6 ]
  %4 = phi i8 [ 0, %0 ], [ %10, %6 ]
  %5 = icmp eq i8 %4, 8
  br i1 %5, label %11, label %6

6:                                                ; preds = %2
  %7 = and i8 %4, 96
  %8 = icmp ne i8 %7, 0
  %9 = zext i1 %8 to i32
  %10 = add nuw nsw i8 %4, 1
  br label %2, !llvm.loop !7

11:                                               ; preds = %2
  store i32 %3, i32* @b, align 4, !tbaa !3
  %12 = icmp eq i32 %3, 0
  br i1 %12, label %14, label %13

13:                                               ; preds = %11
  tail call void (...) @foo() #3
  br label %14

14:                                               ; preds = %13, %11
  ret i32 0
}

; Function Attrs: minsize optsize
declare dso_local void @foo(...) local_unnamed_addr #2

attributes #0 = { minsize mustprogress nofree norecurse nosync nounwind optsize uwtable willreturn writeonly "frame-pointer"="none" "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
attributes #1 = { minsize nounwind optsize uwtable "frame-pointer"="none" "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
attributes #2 = { minsize optsize "frame-pointer"="none" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
attributes #3 = { minsize nounwind optsize }

!llvm.module.flags = !{!0, !1}
!llvm.ident = !{!2}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{i32 7, !"uwtable", i32 1}
!2 = !{!"clang version 14.0.0 (https://github.com/llvm/llvm-project.git f29256a64ac11cf59cea878c8d1ba9537db4f523)"}
!3 = !{!4, !4, i64 0}
!4 = !{!"int", !5, i64 0}
!5 = !{!"omnipotent char", !6, i64 0}
!6 = !{!"Simple C/C++ TBAA"}
!7 = distinct !{!7, !8}
!8 = !{!"llvm.loop.mustprogress"}
```
</p></details>

<details ><summary>Reduced assembly</summary><p>

```asm
main:                                   # @main
	.cfi_startproc
# %bb.0:
	pushq	%rax
	.cfi_def_cfa_offset 16
	movl	b(%rip), %eax
	xorl	%ecx, %ecx
.LBB1_1:                                # =>This Inner Loop Header: Depth=1
	cmpb	$8, %cl
	je	.LBB1_3
# %bb.2:                                #   in Loop: Header=BB1_1 Depth=1
	xorl	%eax, %eax
	testb	$96, %cl
	setne	%al
	incb	%cl
	jmp	.LBB1_1
.LBB1_3:
	movl	%eax, b(%rip)
	testl	%eax, %eax
	je	.LBB1_5
# %bb.4:
	xorl	%eax, %eax
	callq	foo
.LBB1_5:
	xorl	%eax, %eax
	popq	%rcx
	.cfi_def_cfa_offset 8
	retq
.Lfunc_end1:
	.size	main, .Lfunc_end1-main
```
</p></details>

------------------------------------------------

`clang-llvmorg-13.0.0 -Oz [-emit-llvm] -S -o /dev/stdout case.c`

<details ><summary>Emitted IR</summary><p>

```ll
; ModuleID = 'case.c'
source_filename = "case.c"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

@b = dso_local local_unnamed_addr global i32 0, align 4

; Function Attrs: minsize mustprogress nofree norecurse nosync nounwind optsize uwtable willreturn writeonly
define dso_local void @e(i32 %0) local_unnamed_addr #0 {
  %2 = trunc i32 %0 to i16
  %3 = icmp ugt i16 %2, 31
  %4 = zext i1 %3 to i32
  store i32 %4, i32* @b, align 4, !tbaa !3
  ret void
}

; Function Attrs: minsize nounwind optsize uwtable
define dso_local i32 @main() local_unnamed_addr #1 {
  store i32 0, i32* @b, align 4, !tbaa !3
  ret i32 0
}

attributes #0 = { minsize mustprogress nofree norecurse nosync nounwind optsize uwtable willreturn writeonly "frame-pointer"="none" "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
attributes #1 = { minsize nounwind optsize uwtable "frame-pointer"="none" "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }

!llvm.module.flags = !{!0, !1}
!llvm.ident = !{!2}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{i32 7, !"uwtable", i32 1}
!2 = !{!"clang version 13.0.0 (https://github.com/llvm/llvm-project.git d7b669b3a30345cfcdb2fde2af6f48aa4b94845d)"}
!3 = !{!4, !4, i64 0}
!4 = !{!"int", !5, i64 0}
!5 = !{!"omnipotent char", !6, i64 0}
!6 = !{!"Simple C/C++ TBAA"}
```
</p></details>

<details ><summary>Reduced assembly</summary><p>

```asm
main:                                   # @main
	.cfi_startproc
# %bb.0:
	andl	$0, b(%rip)
	xorl	%eax, %eax
	retq
.Lfunc_end1:
	.size	main, .Lfunc_end1-main
```
</p></details>

------------------------------------------------

### Bisection
Bisected to: 8fce94f91610bc4614b1fe0bff8f143ea2b54742
Committed by: @rotateright

------------------------------------------------

`clang-8fce94f91610bc4614b1fe0bff8f143ea2b54742 -Oz [-emit-llvm] -S -o /dev/stdout case.c`
<details ><summary>Emitted IR</summary><p>

```ll
; ModuleID = 'case.c'
source_filename = "case.c"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

@b = dso_local local_unnamed_addr global i32 0, align 4

; Function Attrs: minsize mustprogress nofree norecurse nosync nounwind optsize uwtable willreturn writeonly
define dso_local void @e(i32 %0) local_unnamed_addr #0 {
  %2 = and i32 %0, 65504
  %3 = icmp ne i32 %2, 0
  %4 = zext i1 %3 to i32
  store i32 %4, i32* @b, align 4, !tbaa !3
  ret void
}

; Function Attrs: minsize nounwind optsize uwtable
define dso_local i32 @main() local_unnamed_addr #1 {
  %1 = load i32, i32* @b, align 4, !tbaa !3
  br label %2

2:                                                ; preds = %6, %0
  %3 = phi i32 [ %1, %0 ], [ %9, %6 ]
  %4 = phi i8 [ 0, %0 ], [ %10, %6 ]
  %5 = icmp eq i8 %4, 8
  br i1 %5, label %11, label %6

6:                                                ; preds = %2
  %7 = and i8 %4, 96
  %8 = icmp ne i8 %7, 0
  %9 = zext i1 %8 to i32
  %10 = add nuw nsw i8 %4, 1
  br label %2, !llvm.loop !7

11:                                               ; preds = %2
  store i32 %3, i32* @b, align 4, !tbaa !3
  %12 = icmp eq i32 %3, 0
  br i1 %12, label %14, label %13

13:                                               ; preds = %11
  tail call void (...) @foo() #3
  br label %14

14:                                               ; preds = %13, %11
  ret i32 0
}

; Function Attrs: minsize optsize
declare dso_local void @foo(...) local_unnamed_addr #2

attributes #0 = { minsize mustprogress nofree norecurse nosync nounwind optsize uwtable willreturn writeonly "frame-pointer"="none" "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
attributes #1 = { minsize nounwind optsize uwtable "frame-pointer"="none" "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
attributes #2 = { minsize optsize "frame-pointer"="none" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
attributes #3 = { minsize nounwind optsize }

!llvm.module.flags = !{!0, !1}
!llvm.ident = !{!2}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{i32 7, !"uwtable", i32 1}
!2 = !{!"clang version 14.0.0 (https://github.com/llvm/llvm-project.git 8fce94f91610bc4614b1fe0bff8f143ea2b54742)"}
!3 = !{!4, !4, i64 0}
!4 = !{!"int", !5, i64 0}
!5 = !{!"omnipotent char", !6, i64 0}
!6 = !{!"Simple C/C++ TBAA"}
!7 = distinct !{!7, !8}
!8 = !{!"llvm.loop.mustprogress"}
```
</p></details>

<details ><summary>Reduced assembly</summary><p>

```asm
main:                                   # @main
	.cfi_startproc
# %bb.0:
	pushq	%rax
	.cfi_def_cfa_offset 16
	movl	b(%rip), %eax
	xorl	%ecx, %ecx
.LBB1_1:                                # =>This Inner Loop Header: Depth=1
	cmpb	$8, %cl
	je	.LBB1_3
# %bb.2:                                #   in Loop: Header=BB1_1 Depth=1
	xorl	%eax, %eax
	testb	$96, %cl
	setne	%al
	incb	%cl
	jmp	.LBB1_1
.LBB1_3:
	movl	%eax, b(%rip)
	testl	%eax, %eax
	je	.LBB1_5
# %bb.4:
	xorl	%eax, %eax
	callq	foo
.LBB1_5:
	xorl	%eax, %eax
	popq	%rcx
	.cfi_def_cfa_offset 8
	retq
.Lfunc_end1:
	.size	main, .Lfunc_end1-main
```
</p></details>

------------------------------------------------

Previous commit: 900cc1a226c8b8512fd3d4af90f9d7e5ecc1e722

`clang-900cc1a226c8b8512fd3d4af90f9d7e5ecc1e722 -Oz [-emit-llvm] -S -o /dev/stdout case.c`

<details ><summary>Emitted IR</summary><p>

```ll
; ModuleID = 'case.c'
source_filename = "case.c"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

@b = dso_local local_unnamed_addr global i32 0, align 4

; Function Attrs: minsize mustprogress nofree norecurse nosync nounwind optsize uwtable willreturn writeonly
define dso_local void @e(i32 %0) local_unnamed_addr #0 {
  %2 = trunc i32 %0 to i16
  %3 = icmp ugt i16 %2, 31
  %4 = zext i1 %3 to i32
  store i32 %4, i32* @b, align 4, !tbaa !3
  ret void
}

; Function Attrs: minsize nounwind optsize uwtable
define dso_local i32 @main() local_unnamed_addr #1 {
  store i32 0, i32* @b, align 4, !tbaa !3
  ret i32 0
}

attributes #0 = { minsize mustprogress nofree norecurse nosync nounwind optsize uwtable willreturn writeonly "frame-pointer"="none" "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }
attributes #1 = { minsize nounwind optsize uwtable "frame-pointer"="none" "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="x86-64" "target-features"="+cx8,+fxsr,+mmx,+sse,+sse2,+x87" "tune-cpu"="generic" }

!llvm.module.flags = !{!0, !1}
!llvm.ident = !{!2}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{i32 7, !"uwtable", i32 1}
!2 = !{!"clang version 14.0.0 (https://github.com/llvm/llvm-project.git 900cc1a226c8b8512fd3d4af90f9d7e5ecc1e722)"}
!3 = !{!4, !4, i64 0}
!4 = !{!"int", !5, i64 0}
!5 = !{!"omnipotent char", !6, i64 0}
!6 = !{!"Simple C/C++ TBAA"}
```
</p></details>

<details ><summary>Reduced assembly</summary><p>

```asm
main:                                   # @main
	.cfi_startproc
# %bb.0:
	andl	$0, b(%rip)
	xorl	%eax, %eax
	retq
.Lfunc_end1:
	.size	main, .Lfunc_end1-main
```
</p></details>

Edit: Added IR, as requested in another of these reports

**Comment by fhahn** [comment]
Still reproduces on current `main`: https://clang.godbolt.org/z/f3vhEPeMs

**Comment by rotateright** [comment]
Looks like it got fixed sometime before the clang 15.0 release.
